src/sentry/search/events/builder/metrics.py

```diff
@@ -693,13 +666,17 @@ def resolve_tag_key(self, value: str) -> int | str | None:
         # prior to resolving it via the indexer
         value = self.column_remapping.get(value, value)
 
-        if self.use_default_tags:
-            if value in self.default_metric_tags:
-                return self.resolve_metric_index(value)
-            else:
-                raise IncompatibleMetricsQuery(f"{value} is not a tag in the metrics dataset")
-        else:
+        # Attempt to resolve tag keys, for the metrics preformance datasets we use an allowlist so ondemand keys don't
+        # overlap. If you don't want this functionality for your dataset set valid_tags to an empty sequence
+        # Ondemand should always resolve the metric_index
+        if (
+            (len(self.valid_tags) > 0 and value in self.valid_tags)
+            or len(self.valid_tags) == 0
+            or self.use_on_demand
```

Comment:
Nit: check the override cases first for better legibility and remove the redundant len > 0 check

```suggestion
            len(self.valid_tags) == 0
            or self.use_on_demand
            or value in self.valid_tags
```